frame/merkle-mountain-range/src/mmr/storage.rs

```diff
@@ -78,32 +81,71 @@ where
 	I: 'static,
 	L: primitives::FullLeaf,
 {
-	fn get_elem(&self, pos: u64) -> mmr_lib::Result<Option<NodeOf<T, I, L>>> {
+	fn get_elem(&self, pos: NodeIndex) -> mmr_lib::Result<Option<NodeOf<T, I, L>>> {
 		Ok(<Nodes<T, I>>::get(pos).map(Node::Hash))
 	}
 
-	fn append(&mut self, pos: u64, elems: Vec<NodeOf<T, I, L>>) -> mmr_lib::Result<()> {
-		let mut leaves = crate::NumberOfLeaves::<T, I>::get();
-		let mut size = crate::mmr::utils::NodesUtils::new(leaves).size();
+	// ? Actually we can move this pruning logic outside
+	// ? and perform the pruning every X blocks
+	// ?
+	// ? commented by Xavier
+	fn append(&mut self, pos: NodeIndex, elems: Vec<NodeOf<T, I, L>>) -> mmr_lib::Result<()> {
+		let leaves = crate::NumberOfLeaves::<T, I>::get();
+		let size = crate::mmr::utils::NodesUtils::new(leaves).size();
+
 		if pos != size {
 			return Err(mmr_lib::Error::InconsistentStore)
 		}
 
-		for elem in elems {
-			// on-chain we only store the hash (even if it's a leaf)
-			<Nodes<T, I>>::insert(size, elem.hash());
+		let diff = |a: &[NodeIndex], b: &[NodeIndex]| -> Vec<NodeIndex> {
+			b.iter().filter(|x| !a.contains(x)).cloned().collect()
+		};
+		let elems = elems
+			.into_iter()
+			.enumerate()
+			.map(|(i, elem)| (size + i as NodeIndex, elem))
+			.collect::<Vec<_>>();
+		let leaves = elems.iter().fold(leaves, |acc, (pos, elem)| {
 			// Indexing API is used to store the full leaf content.
-			let key = Pallet::<T, I>::offchain_key(size);
-			elem.using_encoded(|elem| sp_io::offchain_index::set(&key, elem));
-			size += 1;
+			elem.using_encoded(|elem| {
+				offchain_index::set(&Pallet::<T, I>::offchain_key(*pos), elem)
+			});
 
 			if let Node::Data(..) = elem {
-				leaves += 1;
+				acc + 1
+			} else {
+				acc
 			}
-		}
+		});
```

Comment:
These changes seem unnecessary and imho increase both code complexity and the number of iterations + extra allocation.

Reply:
I don't like mutable variable. IMO this simplify the code. 


But yes, more allocation here.